code/drasil-gool/GOOL/Drasil/LanguageRenderer/Constructors.hs

```diff
@@ -150,15 +151,13 @@ binExpr' = on3StateValues (\b v1 v2 ->
 -- either value passed to the operator is a float, this function preserves that 
 -- type by casting the result to a float.
 binExprNumDbl' :: (RenderSym r) => VSBinOp r -> SValue r -> SValue r -> SValue r
-binExprNumDbl' b' v1' v2' = do
-  b <- b'
-  v1 <- v1'
-  v2 <- v2'
+binExprNumDbl' = on3StateWrapped (\b v1 v2 -> do
```

Comment:
I think I preferred the `do`-based version.

Reply:
I have encountered a problem here. Due to the repetition unless I use on3StateWrapped, it won't pass the Hlint. Here is the repetition
```haskell
-- | Constructs binary expressions like v + w, for some operator + and values v 
-- and w, parenthesizing v and w if needed.
binExpr :: (RenderSym r) => VSBinOp r -> SValue r -> SValue r -> SValue r
binExpr b' v1' v2'= do 
  b <- b'
  v1 <- v1'
  v2 <- v2'
  let exprType = numType (valueType v1) (valueType v2)
      exprRender = binExprRender b v1 v2
  mkExpr (bOpPrec b) exprType exprRender

-- | Constructs binary expressions like pow(v,w), for some operator pow and
-- values v and w
binExpr' :: (RenderSym r) => VSBinOp r -> SValue r -> SValue r -> SValue r
binExpr' b' v1' v2'= do 
  b <- b'
  v1 <- v1'
  v2 <- v2'
  let exprType = numType (valueType v1) (valueType v2)
      exprRender = binOpDocD' (RC.bOp b) (RC.value v1) (RC.value v2)
  mkExpr 9 exprType exprRender

-- | To be used in languages where the binary operator returns a double. If 
-- either value passed to the operator is a float, this function preserves that 
-- type by casting the result to a float.
binExprNumDbl' :: (RenderSym r) => VSBinOp r -> SValue r -> SValue r -> SValue r
binExprNumDbl' b' v1' v2' = do 
  b <- b'
  v1 <- v1'
  v2 <- v2'
  let exprType = numType numType (valueType v1) (valueType v2)
      exprRender = binOpDocD' (RC.bOp b) (RC.value v1) (RC.value v2)
  e <- mkExpr 9 exprType exprRender
  binExprCastFloat t1 t2 e
```

Reply:
You should pull out the common lines into a function.  One for applying `numType` and one for `binOpDocD'`. And then rewrite the above in terms of those.